packages/liveblocks-core/src/room.ts

```diff
@@ -2205,31 +2231,80 @@ export function createRoom<
     storageStatus: eventHub.storageStatus.observable,
   };
 
+  //
+  // NOTE:
+  // The Finite State Machine's .transition() function will eventually be
+  // the central gate keeper of sanity. This function will likely be abstracted
+  // away in a proper FSM abstraction (like a separate class).
+  //
+  // The idea is that, eventually, the handler functions can only be called
+  // from here, and never from anywhere else, so this function can perform its
+  // central gatekeeping.
+  //
```

Comment:
If you squint a bit, you can see the state machine structure start to take shape here. There still is quite some work to be done before it's a real one, but this is the closest to that intended structure I can take the current implementation without starting to introduce changes that aren't solely pure refactorings, so out of scope for this PR.